### ISSUE TITLE
What is best way to load json settings from the server?

### ISSUE THREAD

**Issue by wizardnet972** [angular/angular#9047]
Hi,

What is best way to load json settings from the server before AppComponent loaded? but in angular-way

for example, I want SettingService will call to the server and on AppComponent constructor, I'll get the Service ready after response from the server.

I tried, CanActivated but it's not working on base component...

any ideas?

**Comment by pkozlowski-opensource** [comment]
You would have to fetch those settings outside of ng (ex. using `fetch`) and delay boostrap till you've got the response. In any case GitHub issues are not for support questions: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question

**Comment by CaptainCodeman** [comment]
Nooooo ...

You can do it with `APP_INITIALIZER` ...

Well, I was going to post an example of how to do what is probably a common requirement but WTF, why bother. I guess "don't use the framework" is the answer.

**Comment by pkozlowski-opensource** [comment]
@CaptainCodeman sorry for jumping the gun 

> Well, I was going to post an example of how to do what is probably a common requirement but WTF, why bother.

Well, I would say it makes sense to bother since next time you might be one needing help...

**Comment by CaptainCodeman** [comment]
You can use `APP_INITIALIZER` which will execute a function when the app is initialized and delay what it provides if the function returns a promise. This means the app can be initializing and you can also use the existing services and framework features.

As an example, suppose you have a multi-tenanted solution where the site info relies on the domain name it's being served from. This can be [name].letterpress.com or a custom domain which is matched on the full hostname. We can hide the fact that this is behind a promise by using `APP_INITIALIZER`.

In bootstrap:
`provide(APP_INITIALIZER, { useFactory: (sites:SitesService) => () => sites.load(), deps:[SitesService, HTTP_PROVIDERS], multi: true}),`

sites.service.ts:
```
@Injectable()
export class SitesService {
  public current:Site;

  constructor(private http:Http, private config:Config) { }

  load():Promise<Site> {
    var url:string;
    var pos = location.hostname.lastIndexOf(this.config.rootDomain);
    var url = (pos === -1)
      ? this.config.apiEndpoint + '/sites?host=' + location.hostname
      : this.config.apiEndpoint + '/sites/' + location.hostname.substr(0, pos);
    var promise = this.http.get(url).map(res => res.json()).toPromise();
    promise.then(site => this.current = site);
    return promise;
  }
```

NOTE: `config` is just a custom config class. `rootDomain` would be `'.letterpress.com'` for this example and would allow things like `captaincodeman.letterpress.com`.

Any components and other services can now have `Site` injected into them and use the `.current` property which will be a concrete populated object with no need to wait on any promise within the app.

This approach seemed to cut the startup latency which was otherwise quite noticeable if you were waiting for the large Angular bundle to load and then another http request before the bootstrap even began.

**Comment by zoechi** [comment]
@CaptainCodeman I like that side of yours that makes you share such great information about how to better use Angular2 way more than the side bitching about the Router ;-)

**Comment by noammazuz** [comment]
CaptainCodeman you are the man!

**Comment by mhevery** [comment]
This is issue tracking for the Angular project. Please take support questions to one of our forums: 

- http://stackoverflow.com/questions/tagged/angular2
- https://gitter.im/angular/angular
- https://groups.google.com/forum/#!forum/angular

**Comment by code-tree** [comment]
For anyone stumbling on this later, like me. Some things seem to have changed, such as with the bootstrap process. Here's an updated example implementation:

```typescript
export function init_app(first_service: FirstService, second_service: SecondService){
    // Do initing of services that is required before app loads
    // NOTE: 'useFactory' requires a function (that then returns a promise)
    return () => first_service.load()  // + any other services...
}

@NgModule({
    'providers': [
        FirstService,
        SecondService,
        {
            'provide': APP_INITIALIZER,
            'useFactory': init_app,
            'deps': [FirstService, SecondService],
            'multi': true,
        },
    ],
})
export class AppModule {}
```

**Comment by ihommani** [comment]
Hi all,

What would be the difference between injecting some intelligence in the APP_INITIALIZER and injecting the same intelligence in a **ngOnInit** method of the main component of the application (given that component implemnt OnInit )?

Thx !

**Comment by code-tree** [comment]
AFAIK, The app won't load until the promise returned for APP_INITIALIZER has resolved, where as OnInit is called after the main component has already been inited.

My app displays a progress wheel while loading, and I don't want it to disappear until the app is actually ready to be used, so that's the use case for me.

**Comment by ihommani** [comment]
yeah on a second thought i realized it and currently realizing it in my code :) .
Thank you for the clarification.

**Comment by nelisbijl** [comment]
@CaptainCodeman, I am experiencing that angular 2.4.8 no longer waits for the promises being resolved. Did not see that happen in 2.4.7.
Anyone else experiencing this?

**Comment by zoechi** [comment]
Commenting on closed issues is rather pointless.
A new issue with a Plunker to demonstrate the problem would be a better option. 
I tried it yesterday and it worked with the latest Angular version http://stackoverflow.com/a/42409679/217408

**Comment by paulwalker** [comment]
`Cannot find name 'APP_INITIALIZER'`

Where does this variable come from? Does it still exist?

**Comment by DzmitryShylovich** [comment]
@paulwalker `import {APP_INITIALIZER} from '@angular/core';`

**Comment by paulwalker** [comment]
right, that's what I did eventually I was just thrown off because...(old school coder thought constant was global). 

I do think angular should make a better effort to support pre-app initialization like this. There is a scourge of questions and articles of people dealing with this and it falls into security. The angular app should know the security context before any component starts...without having to subscribe to Observables, etc.

**Comment by paulwalker** [comment]
Additionally, in case anyone comes across this, returning an observable did not work for me with the latest version provided by angular-cli, I had to use toPromise()

**Comment by zoechi** [comment]
@paulwalker Did you complete the observable? You could try `.take(1)`

**Comment by DzmitryShylovich** [comment]
@zoechi it only works with Promises. it's by design

**Comment by zoechi** [comment]
@DzmitryShylovich thanks! Quite weird. I was hoping for some consistency :-/

**Comment by DzmitryShylovich** [comment]
@zoechi I think it was implemented before the observable hype :)

**Comment by mslourens** [comment]
It should be explained in the [API docs](https://angular.io/docs/ts/latest/api/core/index/APP_INITIALIZER-let.html) that the function should return a promise. That wasn't clear for me.

**Comment by ysfjwd** [comment]
thanks

**Comment by dorival** [comment]
`HTTP_PROVIDERS` is now **deprecated** in Angular 5. Use `deps: [SitesService, HttpModule],` instead.
Note that you should include `HttpModule` in your `imports` section only, not in `providers` in this scenario.

**Comment by fizaict** [comment]
@code-tree is it also possible to execute a service file before loading the **core.module.ts**. I am stuck here. Please help.
